File: pruebasProyecto/airport.cpp
#include "airport.h"

using namespace std;

// Obtener la instancia única del aeropuerto

/*
Airport& Airport::getInstance() {
    return instance;
}
*/

//Airport Airport::instance;
void Airport::addAirplane(Airplane* _plane){
    airplaneRegister.push_back(_plane);
}
void Airport::addFlight(Flight* _flight){
    string dest1 = _flight->getDestination();
    string date1 = _flight->getDate();
    pair<string,string> pareja(dest1, date1);
    flightsRegister[pareja] = _flight;
}
void Airport::addGate(Gate* _gate){
    gatesRegister.push_back(_gate);
}
void Airport::addHelicopter(Helicopter* _helicopter){
    helicopterRegister.push_back(_helicopter);
}
void Airport::addJet(PrivateJet* _jet){
    privateJetRegister.push_back(_jet);
}
void Airport::addWorker(Worker* _worker){
    airportCrew.push_back(_worker);
}
void Airport::addPassenger(Passenger* _passenger){
    passengersRegistered.push_back(_passenger);
}

void Airport::assignFlight(Flight *vuelo){
  int i = 0;
  bool flag = true;
  while(flag && i < airplaneRegister.size()){
      if(airplaneRegister[i]->getAssignedFlights()< 3){
          airplaneRegister[i]->addFlight(vuelo);
          vuelo->setAirplane((airplaneRegister[i]));
          flag = false;
          cout << "El vuelo ha sido asignado al avion identificado con el codigo " << airplaneRegister[i]->getId() << endl;
      }
      i += 1;
  }

  if(flag){
    cout << "No hay aviones disponibles, pero eso no es problema Cristiano Ronaldo te patrocina un nuevo avion" << endl << "Crea tu avion..." << endl; 
    Airport::getInstance().createAirplane();

  }
}
void Airport:: buyFlight(string destination, string date,Passenger* _passenger){
    pair<string, string> infoRequiered(destination, date);
    map<pair<string, string>, Flight*>::iterator it = flightsRegister.find(infoRequiered);
    if(it != flightsRegister.end()){
        Flight* vuelo = flightsRegister[infoRequiered];
        int seats = vuelo->getAvailableSeats();
        if(seats <= 0)
            cout << "El vuelo para el destino y fecha seleccionada, no se encuentra disponible" << endl;
        else{
            cout << "Su reserva ha sido exitosa" << endl;
            vuelo->setAvailableSeats(seats - 1);
            vuelo->setPassengersRegistered(_passenger);
        }
    }
    else
        cout << "El vuelo no existe para ese destino o fecha" << endl;
}

void Airport::createAirplane(){
    Airplane* airplane = new Airplane;

    string temp;

    cout << "Digite la marca: ";
    cin >> temp;
    airplane->setBrand(temp);
    cout << endl;

    cout << "Digite el modelo: ";
    cin >> temp;
    airplane->setModel(temp);
    cout << endl;

    cout << "Digite el ID: ";
    cin >> temp;
    airplane->setId(temp);
    cout << endl;

    cout << "Digite la capacidad: ";
    int capacity;
    cin >> capacity;
    airplane->setCapacity(capacity);
    cout << endl;

    cout << "Digite la velocidad maxima: ";
    int maxSpeed;
    cin >> maxSpeed;
    airplane->setMaxSpeed(maxSpeed);
    cout << endl;

    cout << "Digite la autonomia (distancia maxima que la aeronave puede recorrer sin recargar combustible): ";
    int autonomy;
    cin >> autonomy;
    airplane->setAutonomy(autonomy);
    cout << endl;

    cout << "Digite el ano: ";
    int year;
    cin >> year;
    airplane->setYear(year);
    cout << endl;

    cout << "Digite la condicion: ";
    cin >> temp;
    airplane->setCondition(temp);
    cout << endl;

    cout << "Digite la ubicacion: ";
    cin >> temp;
    airplane->setUbication(temp);
    cout << endl;

    cout << "Digite la altitud maxima: ";
    int maxAltitude;
    cin >> maxAltitude;
    airplane->setMaxAltitude(maxAltitude);
    cout << endl;

    cout << "Digite la cantidad de motores: ";
    int engineAmount;
    cin >> engineAmount;
    airplane->setEngineAmount(engineAmount);
    cout << endl;

    cout << "Digite la categoria: ";
    cin >> temp;
    airplane->setCategory(temp);
    cout << endl;

    airplane->setAvailability(true);

    Airport::getInstance().addAirplane(airplane);
}

void Airport::createFlight(){
    Flight* flight = new Flight;

    string temp;

    cout << "Digite el ID del vuelo: ";
    int id;
    cin >> id;
    flight->setId(id);
    cout << endl;

    cout << "Digite la fecha del vuelo: ";
    cin >> temp;
    flight->setDate(temp);
    cout << endl;

    cout << "Digite el origen: ";
    cin >> temp;
    flight->setOrigin(temp);
    cout << endl;

    cout << "Digite el destino: ";
    cin >> temp;
    flight->setDestination(temp);
    cout << endl;

    Airport::getInstance().addFlight(flight);
    Airport::getInstance().assignFlight(flight);
}

void Airport::createGate() {
    Gate* gate = new Gate;

    string temp;

    cout << "Digite la ubicacion de la puerta: ";
    cin >> temp;
    gate->setLocation(temp);
    cout << endl;

    cout << "Digite la hora de abordaje: ";
    cin >> temp;
    gate->setBoardingHour(temp);
    cout << endl;

    Airport::getInstance().addGate(gate);
}

void Airport::createPassenger() {
    Passenger* passenger = new Passenger;

    string temp;

    cout << "Digite el ID del pasajero: ";
    int id;
    cin >> id;
    passenger->setId(id);
    cout << endl;

    cout << "Digite el nombre del pasajero: ";
    cin.ignore();  
    getline(cin, temp);
    passenger->setName(temp);
    cout << endl;

    cout << "Digite el apellido del pasajero: ";
    getline(cin, temp);
    passenger->setLastName(temp);
    cout << endl;

    cout << "Digite la fecha de nacimiento del pasajero (YYYY-MM-DD): ";
    cin >> temp;
    passenger->setBirthdate(temp);
    cout << endl;

    cout << "Digite el genero del pasajero: ";
    cin >> temp;
    passenger->setGender(temp);
    cout << endl;

    cout << "Digite la direccion del pasajero: ";
    cin.ignore(); 
    getline(cin, temp);
    passenger->setAddress(temp);
    cout << endl;

    cout << "Digite el numero de telefono del pasajero: ";
    int phoneNumber;
    cin >> phoneNumber;
    passenger->setPhoneNumber(phoneNumber);
    cout << endl;

    cout << "Digite el correo electronico del pasajero: ";
    cin >> temp;
    passenger->setEmail(temp);
    cout << endl;

    cout << "Digite la nacionalidad del pasajero: ";
    cin >> temp;
    passenger->setNationality(temp);
    cout << endl;

    cout << "Digite la cantidad de equipaje del pasajero: ";
    int baggageAmount;
    cin >> baggageAmount;
    passenger->setBaggageAmount(baggageAmount);
    cout << endl;

    cout << "Digite informacion medica del pasajero: ";
    cin.ignore();  
    getline(cin, temp);
    passenger->setMedicalInformation(temp);
    cout << endl;

    Airport::getInstance().addPassenger(passenger);
}

void Airport::createWorker() {
    Worker* worker = new Worker;

    string temp;

    cout << "Digite el ID del trabajador: ";
    int id;
    cin >> id;
    worker->setId(id);
    cout << endl;

    cout << "Digite el nombre del trabajador: ";
    cin.ignore(); 
    getline(cin, temp);
    worker->setName(temp);
    cout << endl;

    cout << "Digite el apellido del trabajador: ";
    getline(cin, temp);
    worker->setLastName(temp);
    cout << endl;

    cout << "Digite la fecha de nacimiento del trabajador (YYYY-MM-DD): ";
    cin >> temp;
    worker->setBirthdate(temp);
    cout << endl;

    cout << "Digite el genero del trabajador: ";
    cin >> temp;
    worker->setGender(temp);
    cout << endl;

    cout << "Digite la direccion del trabajador: ";
    cin.ignore();  
    getline(cin, temp);
    worker->setAddress(temp);
    cout << endl;

    cout << "Digite el numero de telefono del trabajador: ";
    int phoneNumber;
    cin >> phoneNumber;
    worker->setPhoneNumber(phoneNumber);
    cout << endl;

    cout << "Digite el correo electronico del trabajador: ";
    cin >> temp;
    worker->setEmail(temp);
    cout << endl;

    cout << "Digite la posicion del trabajador: ";
    cin >> temp;
    worker->setPosition(temp);
    cout << endl;

    cout << "Digite los años de experiencia del trabajador: ";
    int yearsOfExperience;
    cin >> yearsOfExperience;
    worker->setYearsOfExperience(yearsOfExperience);
    cout << endl;

    cout << "Digite el maximo de horas diarias del trabajador: ";
    int maxDailyHours;
    cin >> maxDailyHours;
    worker->setMaxDailyHours(maxDailyHours);
    cout << endl;

    Airport::getInstance().addWorker(worker);
}

void Airport::createPrivateJet() {
    PrivateJet* privateJet = new PrivateJet;

    string temp;

    cout << "Digite la marca: ";
    cin >> temp;
    privateJet->setBrand(temp);
    cout << endl;

    cout << "Digite el modelo: ";
    cin >> temp;
    privateJet->setModel(temp);
    cout << endl;

    cout << "Digite el ID: ";
    cin >> temp;
    privateJet->setId(temp);
    cout << endl;

    cout << "Digite la capacidad: ";
    int capacity;
    cin >> capacity;
    privateJet->setCapacity(capacity);
    cout << endl;

    cout << "Digite la velocidad maxima: ";
    int maxSpeed;
    cin >> maxSpeed;
    privateJet->setMaxSpeed(maxSpeed);
    cout << endl;

    cout << "Digite la autonomia (distancia maxima que la aeronave puede recorrer sin recargar combustible): ";
    int autonomy;
    cin >> autonomy;
    privateJet->setAutonomy(autonomy);
    cout << endl;

    cout << "Digite el ano: ";
    int year;
    cin >> year;
    privateJet->setYear(year);
    cout << endl;

    cout << "Digite la condicion: ";
    cin >> temp;
    privateJet->setCondition(temp);
    cout << endl;

    cout << "Digite la ubicacion: ";
    cin >> temp;
    privateJet->setUbication(temp);
    cout << endl;


    privateJet->setAvailability(true);

    Airport::getInstance().addJet(privateJet);
}


void Airport::createHelicopter() {
    Helicopter* helicopter = new Helicopter;

    string temp;

    cout << "Digite la marca: ";
    cin >> temp;
    helicopter->setBrand(temp);
    cout << endl;

    cout << "Digite el modelo: ";
    cin >> temp;
    helicopter->setModel(temp);
    cout << endl;

    cout << "Digite el ID: ";
    cin >> temp;
    helicopter->setId(temp);
    cout << endl;

    cout << "Digite la capacidad: ";
    int capacity;
    cin >> capacity;
    helicopter->setCapacity(capacity);
    cout << endl;

    cout << "Digite la velocidad máxima: ";
    int maxSpeed;
    cin >> maxSpeed;
    helicopter->setMaxSpeed(maxSpeed);
    cout << endl;

    cout << "Digite la autonomía: ";
    int autonomy;
    cin >> autonomy;
    helicopter->setAutonomy(autonomy);
    cout << endl;

    cout << "Digite el año: ";
    int year;
    cin >> year;
    helicopter->setYear(year);
    cout << endl;

    cout << "Digite la condición: ";
    cin >> temp;
    helicopter->setCondition(temp);
    cout << endl;

    cout << "Digite la ubicación: ";
    cin >> temp;
    helicopter->setUbication(temp);
    cout << endl;

    cout << "Digite la cantidad de motores: ";
    int engineAmount;
    cin >> engineAmount;
    helicopter->setEngineAmount(engineAmount);
    cout << endl;

    cout << "Digite la capacidad de elevación: ";
    int elevationCapacity;
    cin >> elevationCapacity;
    helicopter->setElevationCapacity(elevationCapacity);
    cout << endl;


    cout << "Digite el uso del helicóptero: ";
    cin >> temp;
    helicopter->setUse(temp);
    cout << endl;

    helicopter->setAvailability(true);
    

    Airport::getInstance().addHelicopter(helicopter);
}

Passenger* Airport::getPassenger(int n){
    return passengersRegistered[n];
}

int Airport::getPassengersRegisterSize(){
    return passengersRegistered.size();
}

void Airport::printAirplanes() const {
    cout << "Airplanes Registered:" << endl;
    for (int i = 0; i < airplaneRegister.size(); ++i) {
        cout << "Airplane " << i + 1 << ":" << endl;
        airplaneRegister[i]->print();
    }
}

void Airport::printHelicopters() const {
    cout << "Helicopters Registered:" << endl;
    for (int i = 0; i < helicopterRegister.size(); ++i) {
        cout << "Helicopter " << i + 1 << ":" << endl;
        helicopterRegister[i]->print();
    }
}

void Airport::printPrivateJets() const {
    cout << "Private Jets Registered:" << endl;
    for (int i = 0; i < privateJetRegister.size(); ++i) {
        cout << "Private Jet " << i + 1 << ":" << endl;
        privateJetRegister[i]->print();
    }
}

File: proyectoPython/airport.py

class Airport:
    def __init__(self):
        self.flights_register = {}
        self.gates_register = []
        self.airplane_register = []
        self.helicopter_register = []
        self.private_jet_register = []
        self.airport_crew = []
        self.passengers_registered = []

    @staticmethod
    def get_instance():
        if not hasattr(Airport, "_instance"):
            Airport._instance = Airport()
        return Airport._instance

    def add_flight(self, flight):
        key = (flight.destination, flight.date)
        self.flights_register[key] = flight

    def add_gate(self, gate):
        self.gates_register.append(gate)

    def add_airplane(self, airplane):
        self.airplane_register.append(airplane)

    def add_helicopter(self, helicopter):
        self.helicopter_register.append(helicopter)

    def add_jet(self, private_jet):
        self.private_jet_register.append(private_jet)

    def add_worker(self, worker):
        self.airport_crew.append(worker)

    def add_passenger(self, passenger):
        self.passengers_registered.append(passenger)

    def create_airplane(self):
        airplane = Airplane()
        try:
            brand = input("Digite la marca: ")
            model = input("Digite el modelo: ")
            id = input("Digite el ID: ")
            capacity = int(input("Digite la capacidad: "))
            max_speed = int(input("Digite la velocidad maxima: "))
            autonomy = int(input("Digite la autonomia: "))
            year = int(input("Digite el ano: "))
            condition = input("Digite la condicion: ")
            ubication = input("Digite la ubicacion: ")
            max_altitude = int(input("Digite la altitud maxima: "))
            engine_amount = int(input("Digite la cantidad de motores: "))
            category = input("Digite la categoria: ")
            airplane.set_brand(brand)
            airplane.set_model(model)
            airplane.set_id(id)
            airplane.set_capacity(capacity)
            airplane.set_max_speed(max_speed)
            airplane.set_autonomy(autonomy)
            airplane.set_year(year)
            airplane.set_condition(condition)
            airplane.set_ubication(ubication)
            airplane.set_max_altitude(max_altitude)
            airplane.set_engine_amount(engine_amount)
            airplane.set_category(category)
            airplane.set_availability(True)
            self.add_airplane(airplane)
        except ValueError:
            print("Error: Asegúrate de ingresar números enteros para la capacidad, velocidad máxima, autonomía, año, altitud máxima, cantidad de motores.")
            del airplane

    def create_flight(self):
        flight = Flight()
        try:
            id = int(input("Digite el ID del vuelo: "))
            date = input("Digite la fecha del vuelo: ")
            origin = input("Digite el origen: ")
            destination = input("Digite el destino: ")
            flight.set_id(id)
            flight.set_date(date)
            flight.set_origin(origin)
            flight.set_destination(destination)
            self.add_flight(flight)
        except ValueError:
            print("Error: Asegúrate de ingresar un número entero para el ID del vuelo.")
            del flight

    def create_gate(self):
        gate = Gate()
        location = input("Digite la ubicacion de la puerta: ")
        boarding_hour = input("Digite la hora de abordaje: ")
        gate.set_location(location)
        gate.set_boarding_hour(boarding_hour)
        self.add_gate(gate)

    def create_passenger(self):
        passenger = Passenger()
        try:
            id = int(input("Digite el ID del pasajero: "))
            name = input("Digite el nombre del pasajero: ")
            last_name = input("Digite el apellido del pasajero: ")
            birthdate = input("Digite la fecha de nacimiento del pasajero (YYYY-MM-DD): ")
            gender = input("Digite el género del pasajero: ")
            address = input("Digite la dirección del pasajero: ")
            phone_number = int(input("Digite el número de teléfono del pasajero: "))
            email = input("Digite el correo electrónico del pasajero: ")
            nationality = input("Digite la nacionalidad del pasajero: ")
            baggage_amount = int(input("Digite la cantidad de equipaje del pasajero: "))
            medical_information = input("Digite información médica del pasajero: ")
            passenger.set_id(id)
            passenger.set_name(name)
            passenger.set_last_name(last_name)
            passenger.set_birthdate(birthdate)
            passenger.set_gender(gender)
            passenger.set_address(address)
            passenger.set_phone_number(phone_number)
            passenger.set_email(email)
            passenger.set_nationality(nationality)
            passenger.set_baggage_amount(baggage_amount)
            passenger.set_medical_information(medical_information)
            self.add_passenger(passenger)
        except ValueError:
            print("Error: Asegúrate de ingresar un número entero para el ID del pasajero y el número de teléfono.")
            del passenger

    def create_worker(self):
        worker = Worker()
        try:
            id = int(input("Digite el ID del trabajador: "))
            name = input("Digite el nombre del trabajador: ")
            last_name = input("Digite el apellido del trabajador: ")
            birthdate = input("Digite la fecha de nacimiento del trabajador (YYYY-MM-DD): ")
            gender = input("Digite el género del trabajador: ")
            address = input("Digite la dirección del trabajador: ")
            phone_number = int(input("Digite el número de teléfono del trabajador: "))
            email = input("Digite el correo electrónico del trabajador: ")
            position = input("Digite la posición del trabajador: ")
            years_of_experience = int(input("Digite los años de experiencia del trabajador: "))
            max_daily_hours = int(input("Digite el máximo de horas diarias del trabajador: "))
            worker.set_id(id)
            worker.set_name(name)
            worker.set_last_name(last_name)
            worker.set_birthdate(birthdate)
            worker.set_gender(gender)
            worker.set_address(address)
            worker.set_phone_number(phone_number)
            worker.set_email(email)
            worker.set_position(position)
            worker.set_years_of_experience(years_of_experience)
            worker.set_max_daily_hours(max_daily_hours)
            self.add_worker(worker)
        except ValueError:
            print("Error: Asegúrate de ingresar un número entero para el ID del trabajador, número de teléfono, años de experiencia y máximo de horas diarias.")
            del worker

    def create_private_jet(self):
        private_jet = PrivateJet()
        brand = input("Digite la marca: ")
        model = input("Digite el modelo: ")
        id = input("Digite el ID: ")
        capacity = int(input("Digite la capacidad: "))
        max_speed = int(input("Digite la velocidad máxima: "))
        autonomy = int(input("Digite la autonomía: "))
        year = int(input("Digite el año: "))
        condition = input("Digite la condición: ")
        ubication = input("Digite la ubicación: ")
        private_jet.set_brand(brand)
        private_jet.set_model(model)
        private_jet.set_id(id)
        private_jet.set_capacity(capacity)
        private_jet.set_max_speed(max_speed)
        private_jet.set_autonomy(autonomy)
        private_jet.set_year(year)
        private_jet.set_condition(condition)
        private_jet.set_ubication(ubication)
        private_jet.set_availability(True)
        self.add_jet(private_jet)

    def create_helicopter(self):
        helicopter = Helicopter()
        brand = input("Digite la marca: ")
        model = input("Digite el modelo: ")
        id = input("Digite el ID: ")
        capacity = int(input("Digite la capacidad: "))
        max_speed = int(input("Digite la velocidad máxima: "))
        autonomy = int(input("Digite la autonomía: "))
        year = int(input("Digite el año: "))
        condition = input("Digite la condición: ")
        ubication = input("Digite la ubicación: ")
        engine_amount = int(input("Digite la cantidad de motores: "))
        elevation_capacity = int(input("Digite la capacidad de elevación: "))
        use = input("Digite el uso del helicóptero: ")
        helicopter.set_brand(brand)
        helicopter.set_model(model)
        helicopter.set_id(id)
        helicopter.set_capacity(capacity)
        helicopter.set_max_speed(max_speed)
        helicopter.set_autonomy(autonomy)
        helicopter.set_year(year)
        helicopter.set_condition(condition)
        helicopter.set_ubication(ubication)
        helicopter.set_engine_amount(engine_amount)
        helicopter.set_elevation_capacity(elevation_capacity)
        helicopter.set_use(use)
        helicopter.set_availability(True)
        self.add_helicopter(helicopter)

    def assign_flight(self, flight):
        i = 0
        flag = True
        while flag and i < len(self.airplane_register):
            if self.airplane_register[i].get_assigned_flights() < 3:
                self.airplane_register[i].add_flight(flight)
                flight.set_airplane(self.airplane_register[i])
                flag = False
                print(f"El vuelo ha sido asignado al avión identificado con el código {self.airplane_register[i].get_id()}")
            i += 1
        if flag:
            print("No hay aviones disponibles, pero eso no es problema Cristiano Ronaldo te patrocina un nuevo avión\nCrea tu avión...")
            self.create_airplane()

    def buy_flight(self, destination, date, passenger):
        flight = self.flights_register.get((destination, date))
        if flight:
            seats = flight.get_available_seats()
            if seats <= 0:
                print("El vuelo para el destino y fecha seleccionada, no se encuentra disponible")
            else:
                print("Su reserva ha sido exitosa")
                flight.set_available_seats(seats - 1)
                flight.set_passengers_registered(passenger)
        else:
            print("El vuelo no existe para ese destino o fecha")

    def get_passenger(self, n):
        return self.passengers_registered[n]

    def get_passengers_register_size(self):
        return len(self.passengers_registered)

    def get_airplane(self):
        return self.airplane_register

    def get_helicopter(self):
        return self.helicopter_register

    def get_private_jet(self):
        return self.private_jet_register